### PR TITLE
Mentions

### PR DESCRIPTION
**Pull request by adrianq**

### CODE REVIEW

**Comment by adrianq** [comment]
Closes EyeSeeTea/dhis2-core#18

File: src/manager/AppManager.js [hunk 1]
```diff
@@ -232,11 +235,110 @@ AppManager.prototype.init = function(callbackFn) {
...
+            type: 'json',
+            params: [
+                'fields=id,comments[mentions]',
+                'filter=comments.user.id:eq:' + t.userAccount.id,
+                'filter=comments.mentions.username:!null',
+                'paging=false'
+            ],
+            success: function (response) {
+                var mentionsUser = {}
+                previousResponse.interpretations.map( interpretation =>
```
Comment (tokland) [inline src/manager/AppManager.js]: map is used to collect results: `xsDoubled = [1, 2, 3].map(x => 2*x)`. To perform side-effects within a loop, it's more declarative forEach: `[1, 2, 3].forEach(n => console.log(n))`

File: src/manager/AppManager.js [hunk 2]
```diff
@@ -232,11 +235,110 @@ AppManager.prototype.init = function(callbackFn) {
...
+                'fields=id,comments[mentions]',
+                'filter=comments.user.id:eq:' + t.userAccount.id,
+                'filter=comments.mentions.username:!null',
+                'paging=false'
+            ],
+            success: function (response) {
+                var mentionsUser = {}
+                previousResponse.interpretations.map( interpretation => 
+                    interpretation.mentions.map(mention => {
+                         if (mention.username in mentionsUser){
```
Comment (tokland) [inline src/manager/AppManager.js]: This saves some lines: `mentionsUser[mention.username] = (mentionsUser[mention.username] || 0) + 1;`

File: src/manager/AppManager.js [hunk 3]
```diff
@@ -232,11 +235,110 @@ AppManager.prototype.init = function(callbackFn) {
...
+                                mentionsUser[mention.username] = 1;
+                             }
+                        }
+                    ))
+                );
+
+                // console.log(mentionsUser);
+                
+                // Sort by mentions
+                mentionsUser = Object.keys(mentionsUser).sort( function(a,b) {
```
Comment (tokland) [inline src/manager/AppManager.js]: It's a bit confusing to reuse the same variable to hold different values. First `mentionUser` is an object, then it's an array of sorted usernames.

File: src/manager/AppManager.js [hunk 4]
```diff
@@ -232,11 +235,110 @@ AppManager.prototype.init = function(callbackFn) {
...
+                    return mentionsUser[b] - mentionsUser[a];
+                });
+
+               
+                // Get real user object (No foreign key in json b)
+                t.mostMentionedUsers = mentionsUser.map(username => {
```
Comment (tokland) [inline src/manager/AppManager.js]: Here it's probably not a problem, we don't have so many users or mentions, but loop within a loop has quadratic behaviour. Tipically you would build an intermediate structure so the operations within the loop are O(1) and the total is a lineal O(n). Something like this (untested):

```
var usersByUsername = {};
t.users.forEach(user => usersByUsername[user.userCredentials.username] = user);
t.mostMentionedUsers = mentionsUser.map(username => usersByUsername[username]);
```
You can create `usersByUsername` is a single line using a `reduce`, but  it's very cryptic, I would not do it. Underscore/lodash have methods (i.e fromPairs()) to make this task very simple.

File: src/ui/EastRegion.js [hunk 5]
```diff
@@ -285,6 +285,96 @@ EastRegion = function(c) {
...
+                                }, label);
+                            }
+                        }
+                    }
+                });
+
+            var potentialUsers = appManager.users
+                .filter(user => user.userCredentials.username.includes(currentMention))
+                .map((user) => {
+                    return {
```
Comment (tokland) [inline src/ui/EastRegion.js]: Code repeated two times, maybe abstracted in a function?

File: src/ui/EastRegion.js [hunk 6]
```diff
@@ -285,6 +285,96 @@ EastRegion = function(c) {
...
+                                    var newText = splitText.join("@") + "@" + user.userCredentials.username;
+                                    component.setValue(newText);
+                                    mentionsPanel.hide();
+                                }, label);
+                            }
+                        }
+                    }
+                });
+
+            if (potentialMostMentionedUsers != null && potentialMostMentionedUsers.length > 0){
```
Comment (tokland) [inline src/ui/EastRegion.js]: `if (potentialMostMentionedUsers != null` In the general case it's better to check simply that is "truthy", like you do in Python: `if (potentialMostMentionedUsers`. This way it covers also `undefined`.

**Review by tokland — CHANGES_REQUESTED** [review]
Some general comments apart from code specifics:

 (using AT to avoid mentioning github users)

- Autocomplete is not implemented in the "Write Interpretation" popup.
- Visual: 1) Start writing a comment, 2) Write "[AT]andr", selector with users appear, 3) Click "Clear interpretation". Problem: The selector dialog is not hidden.
- Visual: Write "Hello [AT]android". Now go back with the cursor, when you get to "Hello", the dialog re-appears. It cannot be closed other than selecting the user already mentionded. It does not allow to write a user mention at this point.

**Comment by adrianq** [comment]
Re the general comments, I have extracted mentionPanel as a module, MentionToolbar, so it can be instantiated both from the EastPanel and the InterpretationWindow. It took some time but it looks much better now.
I am also using the cursor position in the text area. Before this we always consider the cursor to be at the end of the text and this creates some problems as you noticed.
Finally, I have added some event listeners. These listeners handle any click outside the MentionToolbar. If there is a click anywhere we close the mentiontoolbar (I have tried to mimic the Slack approach as much as possible).

File: src/manager/AppManager.js (hunk 1 above)
Comment (adrianq) [reply src/manager/AppManager.js]: Done!

File: src/manager/AppManager.js (hunk 2 above)
Comment (adrianq) [reply src/manager/AppManager.js]: Much better yep. Changed it!

File: src/manager/AppManager.js (hunk 3 above)
Comment (adrianq) [reply src/manager/AppManager.js]: I am now using two different names... TBH I am very bad with naming

File: src/manager/AppManager.js (hunk 4 above)
Comment (adrianq) [reply src/manager/AppManager.js]: Your code was not tested but it works like a charm :) I also improved a bit the comments

File: src/ui/EastRegion.js (hunk 5 above)
Comment (adrianq) [reply src/ui/EastRegion.js]: Extracted as a function. Note this is now part of the module MentionToolbar as it is used by EastRegion and InterpretationWindow

File: src/ui/EastRegion.js (hunk 6 above)
Comment (adrianq) [reply src/ui/EastRegion.js]: Done!

**Pull request closed without merging**